baseclasses/SolverHist.py

```diff
@@ -0,0 +1,210 @@
+"""
+==============================================================================
+BaseClasses: Solver History Class
+==============================================================================
+@File    :   SolverHist.py
+@Date    :   2020/11/30
+@Author  :   Alasdair Christison Gray
+@Description :
+"""
+
+# ==============================================================================
+# Standard Python modules
+# ==============================================================================
+from warnings import warn
+import copy
+import os
+
+# ==============================================================================
+# External Python modules
+# ==============================================================================
+import numpy as np
+
+# ==============================================================================
+# Extension modules
+# ==============================================================================
+
+
+class SolverHist(object):
+    """
+    The SolverHist class stores various useful values during a nonlinear solution in a nested dictionary structure, by
+    default, the history only stores the values from the latest iteration which are used for printing to the terminal.
+    Storing of the full solution history is enabled using the 'writeConvergenceHist' option
+    """
+
+    def __init__(self, options):
+        """
+        Create an empty history dictionary based on the current solver options
+        """
+
+        if options["writeconvergencehist"]:
+            length = int(options["maxiter"] + 1)
+            self.storeValues = True
+        else:
+            length = 1
+            self.storeValues = False
+
+        # Make a local copy of the options
+        self.options = options
+
+        # Define some defaults
+        self.defaults = {"values": {}, "valueFormat": {}}
+        self.defaults["valueFormat"]["float"] = "{:17.11e}"
+        self.defaults["values"]["float"] = np.full(length, np.nan)
+
+        self.defaults["valueFormat"]["int"] = "{:03d}"
+        self.defaults["values"]["int"] = [np.nan] * length
+
+        self.defaults["valueFormat"]["other"] = "{:^40}"
+        self.defaults["values"]["other"] = [None] * length
+
+        self.hist = {"values": {}, "valueFormat": {}, "nameFormat": {}, "varType": {}}
+        self.varsToPrint = []
```

Comment:
Would add some more inline comments in this section

Reply:
Sure, this is just setting up the default printing and storage formats for different variable types. Like you said in another comment, I think we should completely overhaul how this is done so this code will probably be replaced